virtual_ecosystem/core/core_components.py

```diff
@@ -188,148 +196,214 @@ class LayerStructure:
+        canopy layer indices and the surface layer.
+
+        3. The ``flux_layers`` role includes the filled canopy layers and the topsoil
+           layer.
+
+        In addition, the :attr:`.lowest_canopy_filled` attribute provides an array
+        giving the vertical index of the lowest filled canopy layer in each grid cell.
+        It contains ``np.nan`` when there is  no canopy in a grid cell and is initalised
+        as an array of ``np.nan`` values.
+
+    **Getting layer indices**:
+
+        The :attr:`._role_indices_bool` and :attr:`._role_indices_int` attributes
+        contain dictionaries keyed by role name of the boolean or integer indices of the
+        different defined roles. However, all of the role indices should be accessed
+        using the specific instance properties e.g. :attr:`.index_above`.
+
+        Note that the standard index properties like :attr:`.index_above` will return an
+        array index, which extracts a two dimensional slice of the vertical structure.
+        It is sometimes more convenient to extract a 1 dimensional slice across cells,
+        dropping the vertical dimension. This only makes sense for the role layers  that
+        are by definition a single layer thick (``above``, ``surface`` and ``topsoil``),
+        and for these three layers, additional properties (e.g.
+        :attr:`.index_above_scalar`) are defined that will return a scalar index that
+        extracts a one-dimensional slice.
+
+    **Methods overview**:
+
+        * :meth:`.from_template`: this returns an empty DataArray with
+          the standard vertical layer structure and grid cell dimensions used across the
+          Virtual Ecosystem models.
+
+        * :meth:`.set_filled_canopy`: this method is used to update the
+          ``filled_canopy`` role indices, the related ``filled_atmosphere`` and
+          ``flux_layers`` roles, and the :attr:`.lowest_canopy_filled` attribute.
 
     Raises:
         ConfigurationError: If the configuration elements are incorrect for defining
-            the model timing.
+            the layer structure.
     """
 
-    canopy_layers: int = field(init=False)
+    config: InitVar[Config]
+    """A configuration object instance."""
+
+    # These two init arguments could also be accessed directly from the config, but
+    # this allows for the core components flow from Grid and CoreConstants to validate
+    # these values rather than doing it internally.
+    n_cells: InitVar[int]
+    """The number of grid cells in the simulation."""
+    max_depth_of_microbial_activity: float
+    """The maximum soil depth of significant microbial activity."""
+
+    # Attributes populated by __post_init__
+    n_canopy_layers: int = field(init=False)
     """The maximum number of canopy layers."""
-    soil_layers: NDArray[np.float32] = field(init=False)
+    soil_layer_depth: NDArray[np.float32] = field(init=False)
```

Comment:
Would this be better named as `depths` because `depth` makes me think that every layer is of equal thickness